apisix/plugins/google-logging.lua

```diff
@@ -0,0 +1,298 @@
+                project_id = { type = "string" },
+                token_uri = {
+                    type = "string",
+                    default = "https://oauth2.googleapis.com/token"
+                },
+                -- https://developers.google.com/identity/protocols/oauth2/scopes#logging
+                scopes = {
+                    type = "array",
+                    items = {
+                        description = "Google OAuth2 Authorization Scopes",
+                        type = "string",
+                    },
+                    minItems = 1,
+                    uniqueItems = true,
+                    default = {
+                        "https://www.googleapis.com/auth/logging.read",
+                        "https://www.googleapis.com/auth/logging.write",
+                        "https://www.googleapis.com/auth/logging.admin",
+                        "https://www.googleapis.com/auth/cloud-platform"
+                    }
+                },
+                entries_uri = {
+                    type = "string",
+                    default = "https://logging.googleapis.com/v2/entries:write"
+                },
+            },
+            required = { "private_key", "project_id", "token_uri" }
+        },
+        auth_file = { type = "string" },
+        -- https://cloud.google.com/logging/docs/reference/v2/rest/v2/MonitoredResource
+        resource = {
+            type = "object",
+            properties = {
+                type = { type = "string" },
+                labels = { type = "object" }
+            },
+            default = {
+                type = "global"
+            },
+            required = { "type" }
+        },
+        -- https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+        log_id = { type = "string", default = "apisix.apache.org%2Flogs" },
+        max_retry_count = { type = "integer", minimum = 0, default = 0 },
+        retry_delay = { type = "integer", minimum = 0, default = 1 },
+        buffer_duration = { type = "integer", minimum = 1, default = 60 },
+        inactive_timeout = { type = "integer", minimum = 1, default = 10 },
+        batch_max_size = { type = "integer", minimum = 1, default = 100 },
+    },
+    oneOf = {
+        { required = { "auth_config" } },
+        { required = { "auth_file" } },
+    },
+}
+
+
+-- remove stale objects from the memory after timer expires
+local function remove_stale_objects(premature)
+    if premature then
+        return
+    end
+
+    for key, batch in ipairs(buffers) do
+        if #batch.entry_buffer.entries == 0 and #batch.batch_to_process == 0 then
+            core.log.warn("removing batch processor stale object, route id:", tostring(key))
+            buffers[key] = nil
+        end
+    end
+
+    stale_timer_running = false
+end
+
+
+local function send_to_google(oauth, entries)
+    local http_new = http.new()
+    local res, err = http_new:request_uri(oauth.entries_uri, {
+        ssl_verify = false,
```

Comment:
We need to make ssl_verify configurable and default to true.

Reply:
updated